### PR TITLE
channels: support targeting kubernetes versions

### PR DESCRIPTION
**Pull request by justinsb**
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2314)

### CODE REVIEW

File: channels/pkg/channels/addon.go [hunk 1]
```diff
@@ -25,23 +25,50 @@ import (
 	"net/url"
 )
 
+// Addon is a wrapper around a single version of an addon
```
Comment (kris-nova) [inline channels/pkg/channels/addon.go]: Can't say how much I love this enough. Using the type name as the first word, and adding docs.

🎉

File: channels/pkg/channels/addon.go [hunk 2]
```diff
@@ -25,23 +25,50 @@ import (
...
+	Addons map[string]*Addon
+}
+
+func NewAddonMenu() *AddonMenu {
+	return &AddonMenu{
+		Addons: make(map[string]*Addon),
+	}
+}
+
+func (m *AddonMenu) Merge(o *AddonMenu) {
```
Comment (kris-nova) [inline channels/pkg/channels/addon.go]: `s/Merge/MergeAddons/g`? 

Wondering if being a little more explicit here would help?

File: channels/pkg/channels/addons_test.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	for _, g := range grid {
+		actual := g.New.replaces(g.Old)
+		if actual != g.Replaces {
+			t.Errorf("unexpected result from %v -> %v, expect %t.  actual %v", g.Old, g.New, g.Replaces, actual)
+		}
+	}
+}
+
+func s(v string) *string {
```
Comment (kris-nova) [inline channels/pkg/channels/addons_test.go]: Convenience functions! I know you love these, but it creates a questionable layer of complexity..

Whats wrong with just using idiomatic `*` and `&`?

File: docs/addon_manager.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
+## Addons Management
```
Comment (kris-nova) [inline docs/addon_manager.md]: Maybe a pointer in `/README.md` to this file?

File: channels/pkg/channels/addons_test.go (hunk 3 above)
Comment (justinsb) [inline channels/pkg/channels/addons_test.go]: You can't do `&"a"` :-(

File: docs/addon_manager.md (hunk 4 above)
Comment (justinsb) [inline docs/addon_manager.md]: I'm thinking I start to put together a "tao of kops" or something - e.g. feels like kubectl, why versioned manifests, the apiserver, why are so many fields inferred, what is nodeup/protokube etc.  But a separate PR I think...

**Comment by justinsb** [comment]
Thanks @kris-nova :-)

**Pull request merged**